### PR TITLE
Formalise style preference description

### LINKED ISSUE #1256: Consider making a more formal description of planned changes.
First of all, thank you for developing this very useful package. My proposal is related to the project documentation, so not really a feature request. Still I hope addressing it would be useful both to the project contributors and its users.

## Current situation related to planning of style changes.

There are several places in the documentation and outside, that relate to the vision of the future evolution of black. I tried to make a comprehensive collection below.

- The [beta warning](https://black.readthedocs.io/en/latest/installation_and_usage.html#note-this-is-a-beta-product) says:

  > What this means for you is that until the formatter becomes stable, you should expect some formatting to change in the future. That being said, no drastic stylistic changes are planned, mostly responses to bug reports.

- The [contribution guide](https://black.readthedocs.io/en/stable/contributing.html#bird-s-eye-view) warns that proposals to change formatting are unlikely:

  > On the other hand, if your answer is “because I don’t like a particular formatting” then you’re not ready to embrace Black yet. Such changes are unlikely to get accepted. You can still try but prepare to be disappointed.

- The [style issue template](https://github.com/psf/black/blob/master/.github/ISSUE_TEMPLATE/style_issue.md) on the other hand mentions no policy towards which style changes would be considered acceptable.

- In multiple issues, the developers rejected proposals to change the style, bringing up as an argument that ∼2e7 lines out there is blackified, and that it would be disruptive to change the style.

- The [changelog](https://black.readthedocs.io/en/stable/change_log.html) does not seem to describe whether each new version modifies formatting or not.

- Some issues labeled "design" [stay open](https://github.com/psf/black/issues?q=is%3Aissue+is%3Aopen+label%3Adesign) also following the response from the dev team, which seemingly suggests that these ideas were accepted.

- Testing different black versions on the scipy codebase I have obtained the following changeset (for interested, [source here](https://nbviewer.jupyter.org/url/antonakhmerov.org/misc/black_benchmark.ipynb)):

  |         |   files changed |   additions |   deletions |
  |:--------|----------------:|------------:|------------:|
  | 18.3a0  |             601 |       68255 |       53889 |
  | 18.3a1  |             350 |        2143 |        2219 |
  | 18.3a2  |             200 |         980 |        1460 |
  | 18.3a3  |             250 |        1678 |         881 |
  | 18.3a4  |              60 |        1508 |        1233 |
  | 18.4a0  |             525 |       17244 |       16435 |
  | 18.4a1  |             129 |         590 |         381 |
  | 18.4a2  |               1 |           1 |           1 |
  | 18.4a3  |             135 |         762 |         586 |
  | 18.4a4  |               0 |           0 |           0 |
  | 18.5b0  |             265 |        2790 |        3088 |
  | 18.5b1  |             209 |          63 |        1011 |
  | 18.6b0  |               8 |          24 |          72 |
  | 18.6b1  |               0 |           0 |           0 |
  | 18.6b2  |               0 |           0 |           0 |
  | 18.6b3  |               0 |           0 |           0 |
  | 18.6b4  |               0 |           0 |           0 |
  | 18.9b0  |             187 |        4208 |        4033 |
  | 19.3b0  |               0 |           0 |           0 |
  | 19.10b0 |              59 |         570 |         407 |

  It is not clear to me whether the volume of changes diminishes or not. There are open issues (e.g. #1246) requesting features for pinning the version of black.

- I expect that the understanding of what comprises the best style will evolve with time indefinitely. For example, PEP8 [keeps evolving](https://github.com/python/peps/commits/master/pep-0008.txt) despite its overall rather mature age.

  I find this perspective hard to reconcile with the implicitly communicated idea that when black stops being beta, its style won't change anymore.

## Proposal

I believe the discussion of the planned improvements to the black style, and the reliability of black for the users in the long run would be improved by implementing the following changes.

1. Add a design document that reflects the current vision of what the style of black should be, what aspects may not change, and what aspects may change. Document what arguments are considered when deciding to change the default style. List the planned, but not implemented changes, possibly just linking to the issue tracker.
2. Stop using alpha/beta version labeling (since so many projects already rely on black), implement a programmatic way to determine the version of the black style. The minimal approach here would be semver, however I realize that switching from the current versioning scheme to semver is rather hard. This proposal is independent from the first one.

### PR DESCRIPTION
**Pull request by felix-hilden**
### Description
Closes #1256: I reworded our style docs to be more explicit about the style we're aiming for and how it is changed (or isn't).

### Checklist - did you ...

- [x] N/A Add a CHANGELOG entry if necessary?
- [x] N/A Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

File: docs/the_black_code_style/current_style.md [hunk 1]
```diff
@@ -2,10 +2,14 @@
 
 ## Code style
 
-_Black_ reformats entire files in place. Style configuration options are deliberately
-limited and rarely added. It doesn't take previous formatting into account, except for
-the magic trailing comma and preserving newlines. It doesn't reformat blocks that start
-with `# fmt: off` and end with `# fmt: on`, or lines that ends with `# fmt: skip`.
+_Black_ aims for consistency, generality, readability and reducing git diffs. Similar
```
Comment (felix-hilden) [inline docs/the_black_code_style/current_style.md:5]: Any other important aspects to add?

File: docs/the_black_code_style/index.rst [hunk 2]
```diff
@@ -12,6 +12,10 @@ The Black Code Style
 While keeping the style unchanged throughout releases has always been a goal,
 the *Black* code style isn't set in stone. It evolves to accommodate for new features
 in the Python language and, occasionally, in response to user feedback.
+Large-scale style preferences presented in :doc:`current_style` are very unlikely to
+change, but minor style aspects and details might change according to the stability
+policy presented below. Ongoing style considerations are tracked on GitHub with the
+`design <https://github.com/psf/black/labels/T%3A%20design>`_ issue label.
```
Comment (felix-hilden) [inline docs/the_black_code_style/index.rst:18]: Enough for #1256?

Comment (JelleZijlstra) [reply docs/the_black_code_style/index.rst:18]: I think so

File: docs/the_black_code_style/current_style.md [hunk 3]
```diff
@@ -2,10 +2,14 @@
...
+language constructs are formatted with similar rules. Style configuration options are
+deliberately limited and rarely added. Previous formatting isn't taken into account,
+except for the magic trailing comma and preserving newlines. The coding style used by
```
[the remark this answers is not in the document]
Comment (felix-hilden) [reply docs/the_black_code_style/current_style.md:8]: Oh that's right. Maybe we'll go with "minimising dependency on previous formatting" or something like that 👍

**Pull request merged**